enhancements/machine-config/user-data-secret-managed.md

```diff
@@ -0,0 +1,113 @@
+
+Currently the installer creates [pointer Ignition configs][1] which end up as [user][2] [data][3] secrets [referenced by the installer-generated Machine(Sets)][4].  That makes things like the Ignition v2 (spec 3) transition tricky.
+The key point is that the Secret holding the initial stub ignition is completely unmanaged after it's created by the installer.
+In order to upgrade to the new Ignition spec, the MCO needs a way to update the Secret to the new Ignition spec.
+
+
+[1]: https://github.com/openshift/installer/blob/093ca65398fe567bdf63322894496cebbe3d2ade/pkg/asset/ignition/machine/node.go#L30-L36
+[2]: https://github.com/openshift/installer/blob/093ca65398fe567bdf63322894496cebbe3d2ade/pkg/asset/machines/master.go#L161-L170
+[3]: https://github.com/openshift/installer/blob/093ca65398fe567bdf63322894496cebbe3d2ade/pkg/asset/machines/worker.go#L197-L205
+[4]: https://github.com/openshift/installer/blame/093ca65398fe567bdf63322894496cebbe3d2ade/docs/user/aws/install_upi.md#L231-L232
+
+## Motivation
+
+The goal for this enhancement is to have the Secret that contains the stub ignition configuration managed by the MCO. The MCO will later be able to update the stub to a newer version of the spec during an upgrade.
+
+### Goals
+
+The MCO owns the lifecycle of the Secret containing the pointer config. The installer is still responsible to create the initial version of the Secret as that is required to support UPI installation by generating the initial ignition configs until the installer provisions the control plane via MAO and not TF. The MCO takes ownership of the Secret containing the stub ignition once the cluster is up and running. When we'll update to the new Ignition version, the MCO can update the Secret.
+
+The final goal for this enhancement is to be able to manage that Secret from the MCO during the ignition migration to the new spec.
+
+If we don't manage that Secret, its content will be on an unsupported version of Ignition that can't provision new machines joining the cluster.
+
+### Non-Goals
+
+To make this simple and to be able to fit this into the release that will bring Ignition v2, we're not considering per-machine Secret(s) as highlighted here https://github.com/openshift/machine-config-operator/issues/683#issuecomment-493193902 and here https://github.com/openshift/machine-config-operator/issues/683#issuecomment-640602493
+
+We don't want to diverge from the current implementation where the Installer still creates the Secret and stub ignition on installation otherwise we'll have to change far more things in order to keep supporting UPI installation that needs ignition configs before starting to install.
+
+## Proposal
+
+The proposal is pretty simple today:
+
+- The installer keeps creating the Secret that holds the stub ignition config (at any given version) under a new name
+- The secret will change name from `<role>-user-data` to `<role>-user-data-managed` which is going to be v3 at a later time before the release
```

Comment:
why not,

- the installer continues to create the user-provided secret as `<role>-user-data`
- mco can now generate a secret `<role>-user-data-managed` based on the user/installer provided one. _copying if it must or migrating if it can for upgrades.
- the installer changes to using managed one for machinesets as install-time utilizing mco's generation feature.
- the installer continues to use user/installer created pointer for control-plane until we move to mahine-api managed one.

this doesn't break the existence or expectation of the previously created user-data-secret and still allowing us to move/migrate most of our users to generated one.

Reply:
> the installer changes to using managed one for machinesets as install-time utilizing mco's generation feature
Does the installer run bootstrap MC* components before machinesets are generated? (not sure of the ordering there)

> the installer continues to use user/installer created pointer for control-plane until we move to machine-api managed one.
A bit confused on what you mean here, doesn't the MAO inherit the secret from installer directly today?

Reply:
I think Abhinav is bringing up the "installation" part only here - when we migrate the installer to v3, the pointer config `<role>-user-data` will also be v3 so installation still works and the MCO can take over copying that and moving it to `<role>-user-data-managed` for further changes in the future.

Still, I have the same concerns Jerry has around timing, the MCO runs too late to be able to inject the new pointer config name to machinesets